File: dp_tabulation/tree_house_robber/tree_house_robber.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreeHouseRobber:  # LeetCode Q.337.
    def __init__(self):
        self.max_robbery = 0

    def rob_max_properties(self, root: TreeNode | None):
        self.max_robbery -= self.max_robbery  # Reset before DFS.
        self._dfs_money(root)
        return self.max_robbery

    def _dfs_money(self, current_node: TreeNode | None):
        if current_node is None:  # Non-existent node.
            return 0, 0  # MIS two before set cum value & last set cum value.

        # Track two before & last sets' cum values in both sides.
        left_two_before, left_last = self._dfs_money(current_node.left)
        right_two_before, right_last = self._dfs_money(current_node.right)

        total_two_before = left_two_before + right_two_before  # Merge both sides.
        total_last = left_last + right_last

        if total_two_before + current_node.val > total_last:  # Two before set wins.
            total_last = max(total_two_before, total_last)
            total_two_before += current_node.val
            # Swap two sets before returning to parent node.
            total_two_before, total_last = total_last, total_two_before

        else:  # Last set wins.
            total_two_before = total_last

        max_robbery = max(total_two_before, total_last)
        if max_robbery > self.max_robbery:
            self.max_robbery = max_robbery

        return total_two_before, total_last

File: dp_tabulation/tree_house_robber/tree_house_robber.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BinaryTreeHouseRobber // LeetCode Q.337.
{
private:
    pair<int, int> dfs_rob(TreeNode *node)
    { // Pair format: {robbed money w.r.t. node, robbed money w.r.t. children}.
        int children_money = 0;
        int node_money = node->val;

        if (node->left != nullptr)
        {
            pair<int, int> robbed_money_pair = dfs_rob(node->left);
            node_money += robbed_money_pair.second;
            children_money += robbed_money_pair.first;
        }
        if (node->right != nullptr)
        {
            pair<int, int> robbed_money_pair = dfs_rob(node->right);
            node_money += robbed_money_pair.second;
            children_money += robbed_money_pair.first;
        }

        return {max(node_money, children_money), children_money};
    }

public:
    int rob(TreeNode *root)
    {
        return dfs_rob(root).first;
    }
};